presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java

```diff
@@ -184,6 +218,45 @@ private void scheduleRevoking()
         });
     }
 
+    private synchronized void revokeHighMemoryTasks()
+    {
+        if (checkPending.getAndSet(false)) {
+            Collection<SqlTask> sqlTasks = requireNonNull(currentTasksSupplier.get());
+            for (SqlTask task : sqlTasks) {
+                long currentTaskAllocation = task.getTaskInfo().getStats().getRevocableMemoryReservationInBytes();
+
+                if (currentTaskAllocation > maxRevocableMemoryPerTask) {
+                    AtomicLong remainingBytesToRevokeAtomic = new AtomicLong(currentTaskAllocation - maxRevocableMemoryPerTask);
+                    task.getQueryContext().accept(new VoidTraversingQueryContextVisitor<AtomicLong>()
+                    {
+                        @Override
+                        public Void visitQueryContext(QueryContext queryContext, AtomicLong remainingBytesToRevoke)
+                        {
+                            if (remainingBytesToRevoke.get() < 0) {
+                                // exit immediately if no work needs to be done
+                                return null;
+                            }
+                            return super.visitQueryContext(queryContext, remainingBytesToRevoke);
+                        }
+
+                        @Override
+                        public Void visitOperatorContext(OperatorContext operatorContext, AtomicLong remainingBytesToRevoke)
+                        {
+                            if (remainingBytesToRevoke.get() > 0) {
+                                long revokedBytes = operatorContext.requestMemoryRevoking();
+                                if (revokedBytes > 0) {
+                                    remainingBytesToRevoke.addAndGet(-revokedBytes);
+                                    log.debug("taskId=%s: requested revoking %s; remaining %s", task.getTaskInfo().getTaskId(), revokedBytes, remainingBytesToRevoke.get());
+                                }
+                            }
+                            return null;
+                        }
+                    }, remainingBytesToRevokeAtomic);
+                }
+            }
+        }
+    }
+
     private synchronized void runMemoryRevoking()
```

Comment:
We might need to think about renaming this or merging this with above to not confuse people with different spilling strategies.

Reply:
Just moved this to another file to remove confusion about strategies entirely.